elastic/kibana
# [Response Ops][Alerting] Updating invalidate API key task to exclude API keys being used by backfills (merging into feature branch)

Resolves https://github.com/elastic/kibana/issues/174359

## Summary

This PR updates the invalidate API key task to query for API keys used by ad hoc runs. The ad hoc run saved object stores the unencrypted API key ID (a decision that was approved by the kibana-security team) that can be used in the query.

The invalidate API key task queries the `.kibana` index for `api_key_pending_invalidation` tasks and gets the API key ID for key pending invalidation. This PR adds a query with aggregation against the `.kibana` index to look for any `ad_hoc_run_param` saved objects that reference the API key ID and excludes them from invalidation. The invalidation task runs at a scheduled interval so the API key will still be invalidated, just after the ad hoc run task that uses it is done.

## To Verify

1. Change the default config for the invalidation task so it runs faster and then start Kibana

```
diff --git a/x-pack/plugins/alerting/server/config.ts b/x-pack/plugins/alerting/server/config.ts
index a49c393da1e..eb188b747e4 100644
--- a/x-pack/plugins/alerting/server/config.ts
+++ b/x-pack/plugins/alerting/server/config.ts
@@ -65,8 +65,8 @@ export const configSchema = schema.object({
     interval: schema.string({ validate: validateDurationSchema, defaultValue: '60m' }),
   }),
   invalidateApiKeysTask: schema.object({
-    interval: schema.string({ validate: validateDurationSchema, defaultValue: '5m' }),
-    removalDelay: schema.string({ validate: validateDurationSchema, defaultValue: '1h' }),
+    interval: schema.string({ validate: validateDurationSchema, defaultValue: '1m' }),
+    removalDelay: schema.string({ validate: validateDurationSchema, defaultValue: '1m' }),
   }),
```
2. Create 2 detection rules.
3. Schedule a backfill for one of the rules. Use a long start/end time range so the backfill takes a while to run
4. Perform an action on both rules that would invalidate the API key for these rules. You can update the rule, or delete them, or update their API keys.
5. Verify that the backfill task continues running with no errors. Verify that the API key for the rule with no backfill task gets deleted fairly quickly while the API key for the rule with the backfill task sticks around until the backfill is complete.
